pkg/block/fetcher.go

```diff
@@ -574,4 +602,49 @@ func (f *ConsistencyDelayMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta,
 			delete(metas, id)
 		}
 	}
+
+	return nil
+}
+
+// IgnoreDeletionMarkFilter is a MetaFetcher that filters blocks that are marked for deletion; delay duration before current time.
+// The delay duration is to make sure that the replacement block can be fetched before we filter out the old block.
+type IgnoreDeletionMarkFilter struct {
+	logger          log.Logger
+	delay           float64
+	bkt             objstore.BucketReader
+	deletionMarkIds []ulid.ULID
+}
+
+// NewIgnoreDeletionMarkFilter creates IgnoreDeletionMarkFilter.
+func NewIgnoreDeletionMarkFilter(logger log.Logger, bkt objstore.BucketReader, delay float64) *IgnoreDeletionMarkFilter {
+	return &IgnoreDeletionMarkFilter{
+		logger: logger,
+		bkt:    bkt,
+		delay:  delay,
+	}
+}
+
+// DeletionMarkBlocks returns block ids that have deletion-mark.json file.
+func (f *IgnoreDeletionMarkFilter) DeletionMarkBlocks() []ulid.ULID {
+	return f.deletionMarkIds
+}
+
+// Filter filters out blocks that are marked for deletion; delay duration before current time.
+func (f *IgnoreDeletionMarkFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) error {
+	f.deletionMarkIds = []ulid.ULID{}
+
+	for id := range metas {
+		deletionMark, err := metadata.ReadDeletionMark(ctx, f.bkt, f.logger, id.String())
+		if err == metadata.ErrorDeletionMarkNotFound {
+			continue
+		}
+		if deletionMark != nil {
```

Comment:
again. Please do quick fast first. Also why deletionMark can be nil here? I think it cannot. Also you forget to check for any other error than NotFound

Reply:
sure that is a miss on my part, I agree with this.